server/src/main/java/org/elasticsearch/rest/StreamingXContentResponse.java

```diff
@@ -0,0 +1,435 @@
+    private final Releasable onCompletion;
+
+    /**
+     * A listener for the next fragment to become available for transmission after a pause. Completed with the newly-created unique active
+     * {@link AvailableFragmentsResponseBodyPart} within {@link #writeFragment}, and  subscribed to via
+     * {@link AvailableFragmentsResponseBodyPart#getNextPart} when the current {@link AvailableFragmentsResponseBodyPart}
+     * becomes inactive because of a transmission pause.
+     */
+    @Nullable // if the first fragment hasn't been sent yet
+    private SubscribableListener<ChunkedRestResponseBodyPart> nextAvailableFragmentListener;
+
+    /**
+     * A resource to be released when the transmission of the current fragment is complete. Note that we may complete the transmission of
+     * multiple fragments at the same time, if they are all processed by one call to {@link AvailableFragmentsResponseBodyPart#encodeChunk}
+     * and transmitted together.
+     */
+    @Nullable // if not currently sending a fragment
+    private Releasable currentFragmentReleasable;
+
+    /**
+     * @param restChannel  The {@link RestChannel} on which to send the response.
+     * @param params       The {@link ToXContent.Params} to control the serialization.
+     * @param onCompletion A resource which is released when the transmission is complete.
+     */
+    public StreamingXContentResponse(RestChannel restChannel, ToXContent.Params params, Releasable onCompletion) throws IOException {
+        this.restChannel = restChannel;
+        this.params = params;
+        this.onCompletion = onCompletion;
+        this.xContentBuilder = restChannel.newBuilder(
+            restChannel.request().getXContentType(),
+            null,
+            true,
+            Streams.noCloseStream(new OutputStream() {
+                @Override
+                public void write(int b) throws IOException {
+                    assert targetStream != null;
+                    targetStream.write(b);
+                }
+
+                @Override
+                public void write(byte[] b, int off, int len) throws IOException {
+                    assert targetStream != null;
+                    targetStream.write(b, off, len);
+                }
+            })
+        );
+    }
+
+    /**
+     * Close this {@link StreamingXContentResponse}, indicating that there will be no more fragments to send.
+     */
+    @Override
+    public void close() {
+        writeFragment(p -> NO_MORE_FRAGMENTS, () -> {
+            if (isRestResponseFinished.compareAndSet(false, true)) {
+                queueRefs.decRef();
+            }
+        });
+    }
+
+    private Iterator<? extends ToXContent> getChunksIterator(StreamingFragment fragment) {
+        return xContentBuilder.getRestApiVersion() == RestApiVersion.V_7
+            ? fragment.fragment().toXContentChunkedV7(params)
+            : fragment.fragment().toXContentChunked(params);
+    }
```

Comment:
What do you think about moving this method to the StreamingFragment class?

Reply:
Could do, but I don't really see the advantage. `StreamingFragment` doesn't have access to `xContentBuilder.getRestApiVersion()`.